src/data/sym/card.lean

```diff
@@ -36,9 +35,117 @@ Prove the general case of stars and bars.
+output contains `k.succ` elements, the original input can be reconstructed by casting `fin n` back
+to `fin n.succ`. Otherwise, an instance of `fin.last n` has been removed and the input can be
+reconstructed by adding it back. -/
+def decode (n k : ℕ) : sym (fin n) k.succ ⊕ sym (fin n.succ) k → sym (fin n.succ) k.succ
+| (sum.inl x) := x.map fin.cast_succ
+| (sum.inr x) := (fin.last n)::x
+
+/-- As `encode` and `decode` are inverses of each other, `sym (fin n.succ) k.succ` is equivalent
+to `sym (fin n) k.succ ⊕ sym (fin n.succ) k`. -/
+def equivalent (n k : ℕ) : sym (fin n.succ) k.succ ≃ sym (fin n) k.succ ⊕ sym (fin n.succ) k :=
+{ to_fun := encode n k,
+  inv_fun := decode n k,
+  left_inv := λ x, begin
+    rw encode,
+    split_ifs,
+    { cases x,
+      simpa [decode, sym.cons, sym.erase, subtype.mk.inj_eq] using multiset.cons_erase h },
+    { simp only [decode, sym.map_map, subtype.mk.inj_eq, function.comp],
+      convert @sym.map_congr _ _ _ _ id _ _,
+      { rw sym.map_id },
+      { intros a h',
+        cases a,
+        split_ifs,
+        { norm_num at h_1,
+          simp_rw h_1 at h',
+          exact (h h').elim },
+        { norm_num } } },
+  end,
+  right_inv := begin
+    rintro (x|x),
+    { cases x with x hx,
+      rw [decode, encode],
+      split_ifs,
+      { obtain ⟨⟨y, w⟩, v, b⟩ := multiset.mem_map.mp h,
+        rw [fin.cast_succ_mk, fin.last, subtype.mk_eq_mk] at b,
+        rw b at w,
+        exact nat.lt_asymm w w, },
+      { simp only [sym.map_map, function.comp, fin.val_eq_coe, subtype.mk_eq_mk, fin.coe_mk],
+        convert sym.map_congr _,
+        { rw multiset.map_id, },
+        { rintros ⟨g, hg⟩ h',
+          split_ifs with h'',
+          { simp only [fin.coe_mk, fin.cast_succ_mk] at h'',
+            subst g,
+            exact (nat.lt_asymm hg hg).elim, },
+          { refl } } } },
+    { rw [decode, encode],
+      split_ifs,
+      { cases x, simp [sym.cons, sym.erase, multiset.cons_erase] },
+      { apply h,
+        cases x,
+        simpa only [sym.cons] using multiset.mem_cons_self (fin.last n) x_val } }
+  end }
+
+lemma multichoose1_rec (n k : ℕ) :
+  multichoose1 n.succ k.succ = multichoose1 n k.succ + multichoose1 n.succ k :=
+by simpa only [multichoose1, fintype.card_sum.symm] using fintype.card_congr (equivalent n k)
+
+lemma multichoose2_rec (n k : ℕ) :
+  multichoose2 n.succ k.succ = multichoose2 n k.succ + multichoose2 n.succ k :=
+by simp [multichoose2, nat.choose_succ_succ, nat.add_comm, nat.add_succ]
+
+lemma multichoose1_eq_multichoose2 : ∀ (n k : ℕ), multichoose1 n k = multichoose2 n k
+| 0 0 := by simp [multichoose1, multichoose2]
+| 0 (k + 1) := by simp [multichoose1, multichoose2, fintype.card]
+| (n + 1) 0 := by simp [multichoose1, multichoose2]
+| (n + 1) (k + 1) :=
+by simp only [multichoose1_rec, multichoose2_rec, multichoose1_eq_multichoose2 n k.succ,
+  multichoose1_eq_multichoose2 n.succ k]
```

Comment:
Oh what a nice short proof this was. Hopefully we can recover it somehow in the new version with some additional lemmas to help support manipulating equivalences and cardinalities.